Add KenyaEMR OMODs to ozone-kenya distribution

jecihjoy opened the pull request.

ibacher commented on pom.xml:
```diff
@@ -18,6 +18,43 @@
 
   <properties>
     <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+    <openMRSVersion>2.6.2</openMRSVersion>
```
So, it shouldn't be necessary to specify everything here, just the versions of modules that are not included in the O3 RefApp itself, so this is mostly the KenyaEMR custom modules plus the O2 stuff.

ibacher reviewed (commented):
So, in addition to the comment below, I think we need to modify the `maven-dependency-plugin` setup:

```xml
      <plugin>
        <groupId>org.apache.maven.plugins</groupId>
        <artifactId>maven-dependency-plugin</artifactId>
        <version>2.8</version>
        <executions>
          <execution>
            <id>Unpack Ozone</id>
            <phase>generate-resources</phase>
            <goals>
              <goal>unpack-dependencies</goal>
            </goals>
            <configuration>
              <excludeTransitive>true</excludeTransitive>
              <useBaseVersion>true</useBaseVersion>
              <outputDirectory>${project.build.directory}/ozone</outputDirectory>
              <includeArtifactIds>ozone</includeArtifactIds>
            </configuration>
          </execution>
        </executions>
      </plugin>
```

Currently, this just extracts the Ozone base. We also need to add an execution to this that uses the [`copy-dependencies`](https://maven.apache.org/plugins/maven-dependency-plugin/copy-dependencies-mojo.html) goal that excludes the `ozone` dependencies and writes to `${project.build.directory}/${project.artifactId}-${project.version}/distro/binaries/openmrs/modules`.

It will also be necessary to add something like [this execution](https://github.com/ozone-his/ozone/blob/070b3944a8662a6c153dfed2e6d975a5868dcd26/distro/pom.xml#L298) for the Maven ant-run plugin to rename the resulting `*.jar` files to `*.omod`.

I'd also leave out the `scope` for the module dependencies.

rbuisson commented on pom.xml:
```diff
@@ -28,11 +53,161 @@
+      <artifactId>uiframework-omod</artifactId>
+      <version>${uiframeworkVersion}</version>
+    </dependency>
+<!--    <dependency>-->
+<!--      <groupId>org.openmrs.module</groupId>-->
+<!--      <artifactId>spa</artifactId>-->
+<!--      <version>1.0.6</version>-->
+<!--    </dependency>-->
+<!--    <dependency>-->
+<!--      <groupId>org.openmrs.module</groupId>-->
+<!--      <artifactId>reportingcompatibility-api</artifactId>-->
+<!--      <version>${reportingcompatibilityVersion}</version>-->
+<!--    </dependency>-->
```
Shouldn't this be deleted @jecihjoy ?

rbuisson replied:
Explained by @jecihjoy  [here](https://openmrs.slack.com/archives/C02PYQD5D0A/p1706625674892189?thread_ts=1706625272.634089&cid=C02PYQD5D0A).
>  Two OpenMRS modules that are part of kenyaemr distro could not be pulled from public registries (Mekos, OpenMRS and sonatype registries) that is spa module and reportingcompatibility am still investing why

rbuisson reviewed (commented):
(not the actual review, just a little comment)

rbuisson commented on pom.xml:
```diff
@@ -28,11 +53,156 @@
+      <artifactId>uiframework-omod</artifactId>
+      <version>${uiframeworkVersion}</version>
+    </dependency>
+    <dependency>
+     <groupId>org.openmrs.module</groupId>
+      <artifactId>reportingcompatibility-api</artifactId>
+      <version>${reportingcompatibilityVersion}</version>
+    </dependency>
```
This we can remove. Only OMODs can be added as dependencies here.

rbuisson commented on pom.xml:
```diff
@@ -19,6 +19,31 @@
+    <facilityReportingVersion>1.2.0</facilityReportingVersion>
+    <metadatadeployVersion>1.11.0</metadatadeployVersion>
+    <uiframeworkVersion>3.23.0</uiframeworkVersion>
+    <reportingcompatibilityVersion>2.0.9</reportingcompatibilityVersion>
```
This should as well be removed.

jecihjoy replied:
@rbuisson I have found references of the module in KenyaEMR module and I have found a version that's compatible and resolvable so we can just keep it

rbuisson replied:
OK. But you would need to add the `-omod` dependency, right?

rbuisson commented on pom.xml:
```diff
@@ -19,6 +19,31 @@
+    <covid19.version>1.0.4</covid19.version>
+    <afyastat.version>1.2.2</afyastat.version>
+    <prep.version>1.4.2</prep.version>
+    <initializer.version>2.7.0-SNAPSHOT</initializer.version>
```
This will conflict with the Ref App Iniz OMOD which is 2.6.0.
@ibacher  even though that will work to provide 2 versions of the same modules (and OpenMRS will pick up the most recent) it would be cleaner to then exclude the 2.6.0 from what's inherited from Ozone).

You would need to specify this exclusion in the `<id>Exclude some Ozone files</id>` build step.

See an example of how it's done in Ozone Distro Cambodia [here](https://github.com/mekomsolutions/ozone-distro-cambodia/blob/1ac7fe494dfe86cd733c751f5183ee34bd70f95b/base/pom.xml#L141-L142).

jecihjoy replied:
@rbuisson I have made the necessary changes. We have local changes in the module as a result of maintaining a custom appointments module. I have updated these modules to make sure kenyaemr in ozone will run without errors as we work on aligning our appnts functions to the community one.

ibacher replied:
> even though that will work to provide 2 versions of the same modules (and OpenMRS will pick up the most recent) 

Does that work? I'd be surprised if it did... I don't think our module loading code is very sophisticated.

ibacher commented:
So when I try to run this locally, it fails with this message in the logs:

```
ERROR - ModuleFactory.startModules(254) |2024-02-15T19:07:05,460| Module KenyaEMR cannot be started because it requires the following module(s): kenyaemrorderentry 1.6.2,org.kenyahmis.module.appointments 1.6.1 Please install and start these modules first.

```

ibacher commented:
Ok, so at least `kenyaemrorderentry` seems to not be started due to this:

```
ERROR - ModuleFactory.startModules(254) |2024-02-15T19:07:05,381| Module KenyaEMR Order Entry Module cannot be started because it requires the following module(s): appui 1.10.0 Please install and start these modules first.
```

So we will need to add the appui module into the dependencies.

ibacher commented:
There's also one of these:

```
ERROR - Listener.performWebStartOfModules(732) |2024-02-15T19:07:05,751| Unable to refresh the spring application context.  Root Cause was:
org.xml.sax.SAXParseException: cvc-complex-type.3.2.2: Attribute 'local' is not allowed to appear in element 'ref'.
```

ibacher commented:
So, I've added appui and uicommons, but this configuration still doesn't work. In addition to the above, I'm now seeing:

```
WARN - ModuleFactory.startModuleInternal(783) |2024-02-15T19:44:49,814| Error while trying to start module: kenyaemrCharts
org.openmrs.module.ModuleException: Unable to update data model using liquibase.xml. Module: KenyaEMR ETL Management
/* snip */
Caused by: liquibase.exception.MigrationFailedException: Migration failed for change set liquibase.xml::kenyaemrCharts-20171106-1725::aojwang:
     Reason: liquibase.exception.DatabaseException: You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near 'DROP PROCEDURE IF EXISTS sp_populate_etl_covid_19_assessment' at line 53 [Failed SQL: (1064) CREATE PROCEDURE sp_populate_etl_pre_hiv_enrollment_art()
/* snip */
DROP PROCEDURE IF EXISTS sp_populate_etl_covid_19_assessment
```

ibacher commented:
I think we need to get these issues sorted before we can merge this in. This should all be reproducible by checking out this branch and running:

```bash
./mvnw clean package
source target/go-to-scripts-dir.sh
./start.sh
```

And then watching the Docker logs for the OpenMRS server.

jecihjoy commented:
> So, I've added appui and uicommons, but this configuration still doesn't work. In addition to the above, I'm now seeing:
> 
> ```
> WARN - ModuleFactory.startModuleInternal(783) |2024-02-15T19:44:49,814| Error while trying to start module: kenyaemrCharts
> org.openmrs.module.ModuleException: Unable to update data model using liquibase.xml. Module: KenyaEMR ETL Management
> /* snip */
> Caused by: liquibase.exception.MigrationFailedException: Migration failed for change set liquibase.xml::kenyaemrCharts-20171106-1725::aojwang:
>      Reason: liquibase.exception.DatabaseException: You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near 'DROP PROCEDURE IF EXISTS sp_populate_etl_covid_19_assessment' at line 53 [Failed SQL: (1064) CREATE PROCEDURE sp_populate_etl_pre_hiv_enrollment_art()
> /* snip */
> DROP PROCEDURE IF EXISTS sp_populate_etl_covid_19_assessment
> ```

@ibacher lemmi try to push a fixed version of this to nexus then we counter check

ibacher commented:
Technically, MariaDB. I suspect it's something else in the SQL, but I couldn't find where that code comes from at all.

jecihjoy commented:
> ```shell
> ```shell
> ./start.sh
> ```
> 
> 
>     
>       
>     
> 
>       
>     
> 
>     
>   
> ```

We had ran into this issue while upgrading my

> > So, I've added appui and uicommons, but this configuration still doesn't work. In addition to the above, I'm now seeing:
> > ```
> > WARN - ModuleFactory.startModuleInternal(783) |2024-02-15T19:44:49,814| Error while trying to start module: kenyaemrCharts
> > org.openmrs.module.ModuleException: Unable to update data model using liquibase.xml. Module: KenyaEMR ETL Management
> > /* snip */
> > Caused by: liquibase.exception.MigrationFailedException: Migration failed for change set liquibase.xml::kenyaemrCharts-20171106-1725::aojwang:
> >      Reason: liquibase.exception.DatabaseException: You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near 'DROP PROCEDURE IF EXISTS sp_populate_etl_covid_19_assessment' at line 53 [Failed SQL: (1064) CREATE PROCEDURE sp_populate_etl_pre_hiv_enrollment_art()
> > /* snip */
> > DROP PROCEDURE IF EXISTS sp_populate_etl_covid_19_assessment
> > ```
> 
> @ibacher lemmi try to push a fixed version of this to nexus then we counter check

I have reolved this but I have ran into the issue with initializer-appointments dependency
`ERROR - ModuleFactory.startModules(254) |2024-02-15T21:19:40,707| Module Common Reports cannot be started because it requires the following module(s): initializer 2.1.0 Please install and start these modules first.`

I believe we can revert back to the community versions of these modules to avoid blocking the integration work. However trying to revert back to the appointments mdoule in refapp am running into this issue

``` Failure to find org.bahmni.module.appointments:appointments-api:jar:2.0.0-SNAPSHOT in https://mavenrepo.openmrs.org/nexus/content/repositories/public```

I will look more into this issues tomorrow morning

jecihjoy commented:
@ibacher I reverted into community versions of appointments and initializer modules, and removed them from dep list since ideally they should be imported from refapp. Hovewever testing locally am still running into issues, I exec'd into the docker container and all the modules appears to be what's expected. Please help as help.

Here is the error am getting  https://pastebin.com/Aeu65uYZ

ibacher commented:
So that error is a result of the migration to Spring 5 (OpenMRS 2.4.0+). Previously, Spring configuration allowed this syntax:

```xml
<ref local="someBean" />
```

Which was an unfortunately common pattern in our code. With Spring 5, the syntax needs to be:

```xml
<ref bean="someBean" />
```

Unfortunately, the error message isn't really helpful in tracking it down. 

My suspicion here is that we have a module in the dependency list that isn't actually part of the 3.x version of KenyaEMR, but I don't have a good reference point for what modules are included in that. Alternatively, should we be depending on core 2.3.x? If neither of those are an option, we need to manually review each of the KenyaEMR-specific OMODs to figure out which one needs to be fixed (we did this review a while ago for the community-supported ones).

ibacher commented:
@jecihjoy It seems to be [this line](https://github.com/palladiumkenya/openmrs-module-kenyadq/blob/9fa4db51e2248dec7f905e623dc3d597d10ca375/api/src/main/resources/moduleApplicationContext.xml#L100) that's causing the problem. Can we fix and publish a new version of that module or just remove the kenyadq OMOD altogether?

ojwanganto commented:
We'll have to fix it for now although we will retire the module soon. @jecihjoy will provide a fix once she is available.

jecihjoy commented:
> ```shell
> ```shell
> ./start.sh
> ```
> 
> 
>     
>       
>     
> 
>       
>     
> 
>     
>   
> ```

@ibacher thank you for quickly figuring out where the issue was.

jecihjoy commented:
Right now after testing the changes locally am getting this..
`ERROR - ModuleFactory.startModules(254) |2024-02-16T21:08:56,118| Module KenyaEMR cannot be started because it requires the following module(s): org.kenyahmis.module.appointments 2.0.0-SNAPSHOT Please install and start these modules first.
`
yet I removed references to that module in the kenyaemr version we are using here `18.7.3-SNAPSHOT` might you @ibacher have an idea why this is happening?

ojwanganto commented:
Did you remove this from the config.xml?

jecihjoy commented:
@ojwanganto thanks for pointing that out,
This error is still appearing even after updating `kenyadq`

```
INFO - ServiceContext.startRefreshingContext(840) |2024-02-16T21:36:03,961| Refreshing Context
INFO - ServiceContext.doneRefreshingContext(851) |2024-02-16T21:36:04,117| Done refreshing Context
ERROR - Listener.performWebStartOfModules(732) |2024-02-16T21:36:04,117| Unable to refresh the spring application context.  Root Cause was:
org.xml.sax.SAXParseException: cvc-complex-type.3.2.2: Attribute 'local' is not allowed to appear in element 'ref'.
	at org.apache.xerces.util.ErrorHandlerWrapper.createSAXParseException(Unknown Source) ~[xercesImpl-2.12.2.jar:?]
	at org.apache.xerces.util.ErrorHandlerWrapper.error(Unknown Source) ~[xercesImpl-2.12.2.jar:?]

ojwanganto commented:
We may need to check in all our modules. They are quite a number though

ibacher commented:
@ojwanganto I used something like this:

```bash
for f in *.omod; do unzip -p $f moduleApplicationContext.xml webModuleApplicationContext.xml 2>/dev/null | awk -v file=$f 'BEGIN {print file} /ref local=/{print $0}'; done
```

It looks like the version of `appframework` and `metadatadeploy` should also be updated.

This is the output:

```
addresshierarchy-2.17.0.omod
afyastat-1.2.2.omod
appframework-2.13.0.omod
                                <ref local="appFrameworkService"/>
                <ref local="administrativeNotificationService"/>
appointments-1.6.1.omod
appui-1.13.0.omod
attachments-3.0.0.omod
calculation-1.3.0.omod
cohort-3.6.0.omod
commonreports-1.4.0-SNAPSHOT.omod
covid19-1.0.4.omod
facilityreporting-1.2.0.omod
fhir2-1.10.0.omod
groovy-2.2.4.omod
hivtestingservices-1.5.1.omod
htmlformentry-5.1.0.omod
htmlwidgets-1.11.0.omod
idgen-4.10.0.omod
initializer-2.7.0-SNAPSHOT.omod
kenyacore-3.0.1.omod
kenyadq-1.4.0.omod
                                <ref local="kenyaDqService"/>
kenyaemr-18.7.2-SNAPSHOT.omod
kenyaemrCharts-1.6.1.omod
kenyaemrIL-1.3.0.omod
kenyaemrextras-1.4.2.omod
kenyaemrml-1.3.6.omod
kenyaemrorderentry-1.6.2.omod
kenyakeypop-1.3.0.omod
kenyaui-3.0.0.omod
legacyui-1.16.0.omod
metadatadeploy-1.11.0.omod
                                <ref local="metadataDeployService"/>
metadatamapping-1.6.0.omod
metadatasharing-1.9.0.omod
o3forms-2.2.0.omod
openconceptlab-2.1.0.omod
ordertemplates-1.0.2.omod
owa-1.14.0.omod
patientflags-3.0.5.omod
patientsummary-2.2.0.omod
prep-1.4.2.omod
queue-2.0.0.omod
referencedemodata-2.4.0.omod
reporting-1.25.0.omod
reportingcompatibility-2.0.9.omod
reportingrest-1.14.0.omod
serialization.xstream-0.2.16.omod
teleconsultation-2.0.0-20230831.113926-1.omod
uicommons-2.16.0.omod
uiframework-3.23.0.omod
vdot-1.0.0-SNAPSHOT.omod
```

ojwanganto commented:
Brilliant! @jecihjoy already tried to upgrade the module versions. She will provide an update here

jecihjoy commented:
@ibacher I have updated the module versions, in my local instance these modules run and I have even tested with `platform 2.6.4` however in the ozone kenyaemr distro htmfformentry and metadadeploy re failing 
full server error https://pastebin.com/ZGhd2rMG

ibacher commented:
@jecihjoy Looks like that was fixed in [the latest SNAPSHOT version](https://github.com/openmrs/openmrs-module-htmlformentry/commit/28fa7a81f0c34ae17f3a53ed1d9530be3ac88bd4).

jecihjoy commented:
@ibacher @ojwanganto @rbuisson finally KenyaEMR has loaded.
On login session data is not being persisted because of the error below
![Screenshot from 2024-02-20 17-26-38](https://github.com/palladiumkenya/kenyahmis/assets/18227275/0bddb012-2edd-4d66-8033-d7514a3532f4)

Am not sure how this is resolved on a docker container but I think @ibacher will be able to pick up from here
